### PR TITLE
Add Solidity language

### LINKED ISSUE #160: Solidity support
### Check for existing issues

- [X] Completed

### Language

Solidity

### Tree Sitter parser link

https://github.com/JoranHonig/tree-sitter-solidity

### Language server link

https://github.com/qiuxiang/solidity-ls

### Misc notes

_No response_

### LINKED ISSUE #4872: Solidity support
### Check for existing issues

- [X] Completed

### Language

Solidity

### Tree Sitter parser link

https://github.com/JoranHonig/tree-sitter-solidity

### Language server link

https://github.com/qiuxiang/solidity-ls

### Misc notes

_No response_

### PR DESCRIPTION
**Pull request by tomholford**
# Context

This adds support for Solidity to zed

Resolves zed-industries/extensions#160 

# Preview

## Syntax Highlighting

image

## Go To Definition

go-to-definition

## Auto-Formatting

formatting

##  Tooltips

tooltip

### CODE REVIEW

**Comment by tomholford** [comment]
@cia-bot check

**Comment by tomholford** [comment]
@cla-bot check

File: docs/src/languages/solidity.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Solidity
+
+- Tree Sitter: [tree-sitter-solidity](https://github.com/parmanuxyz/tree-sitter-solidity)
+- Language Server: [solidity](https://github.com/nwolverson/solidity-language-server)
```
Comment (stevennevins) [inline docs/src/languages/solidity.md:4]: Is this link for the LSP repo accurate?

Comment (tomholford) [reply docs/src/languages/solidity.md]: Good catch, that was a leftover from copy-pasting another language's markdown file. Just pushed a commit with the correct link. Thanks @stevennevins

Comment (stevennevins) [reply docs/src/languages/solidity.md]: Nice, the one you linked is the LSP I used to use. Have you tried this [one](https://github.com/NomicFoundation/hardhat-vscode/tree/development/server)?

Comment (tomholford) [reply docs/src/languages/solidity.md]: @stevennevins Not familiar with that one until now. Will give it a shot and push a commit shortly.

Comment (tomholford) [reply docs/src/languages/solidity.md]: Okay just switched to Nomic's language server, rebuilt the binary, and tested again. Looks good! Thanks for the tip 👍

Comment (tomholford) [reply docs/src/languages/solidity.md]: Got it, appreciate the feedback. I am coming back to Solidity after some time, so it's helpful to know what the latest preferred tooling is.

**Comment by lutfi-haslab** [comment]
When this feature release?

File: crates/zed/src/languages/solidity/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  "-"
+  "+"
+  "++"
+  "--"
+] @operator
+
+[
+  "delete"
+  "new"
+] @keyword.operator
```
Comment (SomeoneToIgnore) [inline crates/zed/src/languages/solidity/highlights.scm:217]: Let's add all missing newlines in the files around.

**Review by SomeoneToIgnore — COMMENTED** [review]
I'm trying to run this on https://github.com/SunWeb3Sec/DeFiHackLabs (repo chosen arbitrarily) and see almost zero LSP functions:

* completions are very sparse and seem not to allow to complete what the code around uses
image

* no go to definition seem to work at all

* same for renames

* at least, LSP server downloads fine and seem to format the project fine

Have I done something wrong, is it a regular experience with Solidity LSP or can we make it better?
https://github.com/NomicFoundation/hardhat-vscode/tree/development/server states all those features.

**Comment by tomholford** [comment]
@SomeoneToIgnore: Thanks for marking as draft. Quick update: Had to focus on other priorities, but will be able to pivot back to this soon.

At this point I have some of the LSP features working: go to definition, basic type tooltips, auto-formatting, syntax highlighting. I'm going to try to take another crack at getting the advanced features mentioned in the `solidity-language-server` docs working.

I have pushed a wip commit with my progress so far. I am new to this project (and Rust) and could use help if you have a moment. Do you happen to know why the `print_completion_item` function is not printing any output when I run the binary in dev mode?

**Comment by SomeoneToIgnore** [comment]
Great news!
I'm afraid I have no idea why `print_completion_item` does not print, but do you say that non-dev mode works just fine?

**Comment by tomholford** [comment]
> Great news! I'm afraid I have no idea why `print_completion_item` does not print, but do you say that non-dev mode works just fine?

@SomeoneToIgnore I have rebased on `main`, run in both dev and non-dev modes, and the features highlighted in the PR description are working as expected. I think this is a good first cut for Solidity integration with zed. A future iteration could add the advanced features / deeper integration with the Solidity LSP. So, going to re-open this for review.

**Comment by maxdeviant** [comment]
This is going to have some conflicts with #7467, so we're going to land that first.

**Comment by tomholford** [comment]
> This is going to have some conflicts with #7467, so we're going to land that first.

@maxdeviant Thanks for the heads up. Now that #7467 is merged, I rebased this branch on `main` and made the appropriate changes to this diff. Tested locally with both `cargo run` and `cargo run --release` and still working as expected. So, requesting a re-review.

**Review by SomeoneToIgnore — CHANGES_REQUESTED** [review]
I have not noticed any QOL improvements in the project I linked above since the last review:

* zero rename, tooltips or go to definition capabilities still
* while the completion list got bigger, it seems that now it just tries to add every possible thing to the completion list?
E.g. here it allows me to autocomplete every "struct" after the `.` which does not seem right at all:
image

Is there a public project that I can try where something from above works?

**Comment by koloz193** [comment]
> I have not noticed any QOL improvements in the project I linked above since the last review:
> 
> * zero rename, tooltips or go to definition capabilities still
> * while the completion list got bigger, it seems that now it just tries to add every possible thing to the completion list?
>   E.g. here it allows me to autocomplete every "struct" after the `.` which does not seem right at all:
>   image
> 
> Is there a public project that I can try where something from above works?

do you mean a different solidity repo or a different editor with autocomplete? if the latter, i use vscode which doesnt really support great autocomplete (if any at all). i think that even having syntax highlighting and error indication would satisfy most people, although full support is def better

**Comment by SomeoneToIgnore** [comment]
Great point about VSCode, I've tried to use https://marketplace.visualstudio.com/items?itemName=JuanBlanco.solidity extension and found its functionality quite morbid too.

Speaking of 

> error indication

I've noticed that VSCode's plugin is able to show those, yet Zed's does not at all?
image

Also notice that there's hovers shown too, but those seem quite useless information-wise.

------------------

I would still want get down to the bottom of https://github.com/zed-industries/zed/pull/7152#issuecomment-1932588676 comment: 

> and the features highlighted in the PR description are working as expected

I see no "go to definition" and hovers for https://github.com/SunWeb3Sec/DeFiHackLabs but see those in the PR description.
What is wrong here? The project in question? Me trying the wrong definitions for those features?
Before merging this, I would love to experience myself that the features declared in the PR description work.

**Comment by SomeoneToIgnore** [comment]
Thank you, that makes things more comparable indeed!
At least, the formatting functionality and zero error highlights behavior matches now.

Yet, that plugin shows hovers and allows to go to definition in the project I use:
image

and current Zed's version does not, so my questions in the previous message hold.

**Comment by tomholford** [comment]
> Thank you, that makes things more comparable indeed! At least, the formatting functionality and zero error highlights behavior matches now.
> 
> Yet, that plugin shows hovers and allows to go to definition in the project I use: image
> 
> and current PR's version does not, so my questions in the previous message hold.

Thanks for the feedback. Any suggestions on how to proceed from here? Tried [to print debug output](https://github.com/zed-industries/zed/pull/7152#issuecomment-1930313479 ) to the console, but no luck with that.

**Comment by maxbrunsfeld** [comment]
Just a heads up - we are very soon going to be moving most of Zed's built-in languages into installable extensions (https://github.com/zed-industries/zed/issues/7096). Solidity is a major enough language that I'm ok with merging this in the meantime. But once the extension system lands, we will probably remove this code and put it into a new `solidity` repository under the new [zed-extensions](https://github.com/orgs/zed-extensions/repositories) GitHub org. We'll probably add you as a collaborator on that repo when we do it, if you're interested in helping maintain it.

It would be good to understand why some of the LSP features aren't working for @SomeoneToIgnore . @tomholford so just to confirm - the LSP works for you reliably? What Solidity projects are you testing against?

**Comment by tomholford** [comment]
> Just a heads up - we are very soon going to be moving most of Zed's built-in languages into installable extensions (#7096). Solidity is a major enough language that I'm ok with merging this in the meantime. But once the extension system lands, we will probably remove this code and put it into a new `solidity` repository under the new [zed-extensions](https://github.com/orgs/zed-extensions/repositories) GitHub org. We'll probably add you as a collaborator on that repo when we do it, if you're interested in helping maintain it.

Thanks for the heads up. And that sounds good, I'm happy help any way that I can.

> It would be good to understand why some of the LSP features aren't working for @SomeoneToIgnore . @tomholford so just to confirm - the LSP works for you reliably? What Solidity projects are you testing against?

Confirming that the features described in the PR's *Preview* section above WOMM. As a test project, I have been using this:

https://github.com/PopPunkLLC/gaslite-core

**Comment by SomeoneToIgnore** [comment]
I see that things like Java are being added as Zed extensions: https://github.com/zed-industries/extensions/pull/57
There seems to be slightly less nitpicking in reviews for such PRs (at least, I'm not reviewing anything there, so won't complain about anything 🙂 ) as those are considered "standalone" from Zed code.

Maybe, it's worth converting this into the extension and submitting a PR there.

File: crates/zed/src/languages/solidity/tags.scm [hunk 3]
```diff
@@ -0,0 +1,43 @@
+;; Method and Function declarations
+(contract_declaration (_
+    (function_definition
```
Comment (maxbrunsfeld) [inline crates/zed/src/languages/solidity/tags.scm:3]: Zed doesn't use `tags.scm`. Could you remove this one?

**Review by maxbrunsfeld — CHANGES_REQUESTED** [review]

**Comment by tomholford** [comment]
> I see that things like Java are being added as Zed extensions: [zed-industries/extensions#57](https://github.com/zed-industries/extensions/pull/57) There seems to be slightly less nitpicking in reviews for such PRs (at least, I'm not reviewing anything there, so won't complain about anything 🙂 ) as those are considered "standalone" from Zed code.
> 
> Maybe, it's worth converting this into the extension and submitting a PR there.

Good call, going to close this and reopen as an extension.

**Pull request closed without merging**